feat: Implement Form-related components

who0803 opened the pull request:
### 1️⃣ PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
<br>

### 2️⃣ 이슈 번호
ex) close #130
<br>
<br>

### 3️⃣ 변경 사항
- Form 관련 컴포넌트들은 전부 useForm 커스텀 훅을 기반으로 구현
- useForm 훅 사용 필수
- Form 컴포넌트 구현
- Input 컴포넌트 구현
- Textarea 컴포넌트 구현
- ErrorMessage 컴포넌트 구현
<br>
<br>

### 4️⃣ 테스트 결과
![bandicam_2023-02-03_07-31-53-902_AdobeExpress](https://user-images.githubusercontent.com/65334125/216464860-126814ec-b569-4546-b401-e9540fb1e0b0.gif)

who0803 committed:
feat: Implement Form-related components

- Form 관련 컴포넌트들은 전부 useForm 커스텀 훅을 기반으로 구현
- useForm 훅 사용 필수
- Form 컴포넌트 구현
- Input 컴포넌트 구현
- Textarea 컴포넌트 구현
- ErrorMessage 컴포넌트 구현

pereng11 commented on src/components/Auth/Textarea/Textarea.tsx:
```diff
@@ -0,0 +1,76 @@
+        {...props}
+        {...getFieldProps(name)}
+      />
+      {children ? cloneElement(children, { name }) : null}
```
cloneElement는 써본 적 없는 메서드라서 신기하네요
어떤 이유로 사용하는 건가요>?

who0803 replied:
children 컴포넌트에 props를 전달해주기 위해서 사용했습니다.

pereng11 commented on src/components/Auth/Input/Input.tsx:
```diff
@@ -0,0 +1,118 @@
+   * - 작은 사이즈 input (200px) <br>
+   * - label 사용으로 placeholer 사용 안함 <br>
+   */
+  size: 'sm' | 'md' | 'lg' | 'lg2';
```
lg2보다는 xl이 적당하지 않을까요?

who0803 replied:
근데 두개의 컴포넌트의 크기가 같아서 적당한 이름이 생각나지 않아 lg ver2 느낌으로 설정했습니다.
크기가 같아도 xl 괜찮을까요?

pereng11 replied:
앗 제가 코드를 꼼꼼하게 못 본 것 같습니다. size prop에 있길래 당연히 크기가 다른 거라고 생각했네요
lg2를 사용하는 목적이 label을 숨김처리하기 위함이라면 불리언 값을 갖는 prop을 하나 더 만드는 게 더 좋을 것 같습니다

who0803 replied:
모양도 다릅니다!!

pereng11 replied:
size prop은 딱 스타일만 담당하는게 좋을 것 같아서 고민되네요
회의 때 한 번 더 이야기해보고 싶습니다!

who0803 replied:
border-radius를 prop으로 바꾸겠습니다~

who0803 replied:
border : 'rec' | 'round'

pereng11 commented on src/components/Auth/Input/Input.tsx:
```diff
@@ -0,0 +1,118 @@
+   */
+  placeholder?: string;
+  children?: ReactElement<any, string | JSXElementConstructor<any>>;
+};
```
JSXElementConstructor는 어떤 타입인가요?

who0803 replied:
확실한 답변인지는 모르겠지만

clonElement가 jsx엘리먼트를 생성하기 때문에 JSXElementConstructor를 사용했습니다.

pereng11 replied:
제네릭의 인수로 any를 사용할거라면 다른 곳처럼 ReactNode로 사용해도 될 것 같습니다. 가독성이나 타입 허용 범위 측면에서 더 좋을 것 같아요!

who0803 replied:
ReactNode를 사용했을 경우 아래와 같은 에러가 발생하여
ReactElement< any > 로 변경하겠습니다!

![20230206094523](https://user-images.githubusercontent.com/65334125/216857148-7973bc39-2a78-4734-b231-b7452ee9669a.png)

pereng11 commented on src/components/Auth/ErrorMessage/ErrorMessage.tsx:
```diff
@@ -0,0 +1,57 @@
+   * 서버메세지를 나타냅니다.
+   * 서버에러가 있다면 서베 에러메세지를 string으로 넣어주면 됩니다.
+   */
+  serverError?: string | null;
```
serverError를 error와 구분하여 따로 받는 이유가 무엇인가요?

who0803 replied:
우선 error는 useForm 훅의 인자로 넣어줄 객체의 validate 함수를 통해 결정되는 validation 에러를 나타내고

반면에 serverError는 useForm 훅의 인자로 넣어줄 객체의 onSubmit 함수를 통해 발생할 수도 있는 서버측 에러(ex 아이디 중복)를 나타냅니다.

그리고 error는 데이터 구조가 input의 name을 키값으로 갖고 error 메세지를 밸류로 갖는 형식인 반면에
serverError는 에러메세지 string 값입니다.

제가 후에 pr 날릴 Signin, Signup 페이지를 참고하시면 될 것같습니다!

who0803 replied:
그리고 ErrorMessage 컴포넌트는 반드시 Input 컴포넌트와 짝을 맞춰서 validation 용도로만 사용해야만 하는 것이 아니고

ErrorMessage 컴포넌트 단독으로 에러메세지를 표현할 수 있기 때문에 serverError props를 따로 마련했습니다.

예를 들어 "로그인에 실패하였습니다." 라는 에러메세지 같은 경우는 ErrorMessage 컴포넌트 단독으로 사용 가능합니다.

pereng11 replied:
Input과 관계없이 독립적으로 사용할 수 있는 컴포넌트를 의도하신 거라면 오히려 에러의 종류를 알 필요 없이 내려주는 텍스트만 그리는 게 더 적합하지 않을까하는 생각이 듭니다.
 ErrorMessage 컴포넌트는 Atom단위의 컴포넌트에 가까운데, 에러의 종류가 무엇인지, input에 연결되었는지를 ErrorMessage컴포넌트에서 처리하면 범용성이 더 떨어질 것 같아서요.

같은 이유로 touched와 name prop을 받는 이유가 렌더링 여부를 컴포넌트 내부에서 결정하기 위함이라면 ,플래그 변수로 사용할 prop을 대신 만들고 그 값은 외부에서 결정해주는 것이 관심사 분리 측면에서 더 좋을 것 같은데 어떻게 생각하시나요?

who0803 replied:
저도 재영님 의견처럼 ErrorMessage 컴포넌트를 아토믹하게 만드는 것도 좋은 생각인 것같습니다!

하지만 ErrorMessage 컴포넌트를 만들 때 의도한 것은 단순히 에러메세지를 보여주는 아토믹 컴포넌트가 아닌 리액트 라이브러리인 react-hook-form 처럼 input 값의 validation에 따라 에러메세지의 내용과 표시 유무가 정해지는 것이 였습니다.

그렇지만 저도 serverError prop을 error prop과 통일하지 못한 부분에 대해 아쉬움이 있었고 재영님께서 의견주신 내용에 동의하기 때문에 기존 ErrorMessage 컴포넌트에서 serverError 로직 부분을 제외하고 에러메세지(string)를 prop로 내려주는 아토믹 Message 컴포넌트를 만들어서 이 아토믹한 Message 컴포넌트를 ErrorMessage 컴포넌트에서 사용하도록 수정하려 하는데 어떻게 생각하시나요?

그렇게 된다면 제가 의도한 validation 체크가 가능한 컴포넌트와 재영님께서 의견을 주신 범용성있는 컴포넌트도 가능하다고 생각합니다!

who0803 replied:
![20230206112937](https://user-images.githubusercontent.com/65334125/216869679-00742527-eb25-47c2-ac51-a5a087337e5c.png)
![20230206112925](https://user-images.githubusercontent.com/65334125/216869686-de572fe8-945f-43a5-9476-2e69896c39e7.png)

pereng11 replied:
넵 좋습니다!! 
컴포넌트 구성은 좋은데 혹시 이름 조금만 수정 괜찮으신가요?
둘 다 에러메시지로 쓸 컴포넌트라면 Error가 둘 다 들어가는게 컴포넌트를 이해하기 더 좋을 것 같습니다!

who0803 replied:
Message -> ErrorMessage
ErrorMessage -> ValidationErrorMessage

pereng11 commented on src/components/Auth/ErrorMessage/ErrorMessage.tsx:
```diff
@@ -0,0 +1,57 @@
+  if (serverError) {
+    return <p className={`${styles.error} ${className}`}>{serverError}</p>;
+  }
+  if (!name || !touched || !errors) return null;
+  if (!touched[name] || !errors[name]) {
+    return null;
+  }
```
상태값에 따라 선택적으로 그려지는 컴포넌트를 구현하는 방법을 두가지로 나누어볼 수 있을 것 같습니다.
1. 상태를 직접 관리하고있는 컨테이너 컴포넌트에서 옵셔널 렌더링을 한다
2. 해당 컴포넌트 내부에서 props에 따라 null을 반환한다
개인적으로는 1번 방법을 선호하는 편이라, 2번 방법으로 구현하신 이유가 무엇인지 궁금합니다.

who0803 replied:
저도 보통은 1번 방식으로 구현하지만 

에러메세지 컴포넌트 같은 경우에 1번 방식대로 구현하게 된다면 컴포넌트를 사용하는 곳에서 마다

옵셔널 렌더링을 컴포넌트를 사용하는 사용자가 구현해줘야 하기 때문에

사용자 경험을 위해서 2번 방식으로 구현했습니다.

pereng11 commented on src/components/Auth/Input/Input.tsx:
```diff
@@ -0,0 +1,118 @@
+      <label
+        htmlFor={inputId}
+        className={`${styles.label} ${size === 'lg2' ? 'srOnly' : ''} ${
+          size !== 'lg' ? '' : value ? styles.top : 'qwe'
```
1. qwe는 무엇인가요?
2. size가 lg이고 value가 있을 때만 styles.top을 주고싶은 거라면 아래와 같은 방법은 어떤가요? 옵셔널 체이닝이 중첩되면 가독성이 떨어지는 것 같더라구요
```javascript
{size === 'lg' && value ? styles.top : ''}
```

who0803 replied:
1. qwe는 개발하면서 class이름 잘들어갔는지 확인하다가 안뺀거 같습니다...  당장 수정할겠습니다!
2. 너무 좋습니다! 제가 왜 저렇게 했을까요

pereng11 reviewed (commented):
스토리북 파일까지 꼼꼼하게 만들어주셨네요 👍 👍 

질문 몇 개 추가합니다!

1. Auth폴더 내부의 Input, Textarea, ErrorMessage 컴포넌트
  - Auth관련한 곳에서만 사용할 컴포넌트인가요? 구현 내용을 보면 공통 컴포넌트에 가까워보여서 질문드립니다
  - getFieldProps 반환값이 아닌 함수 자체를 props로 받아오는 이유가 무엇인가요? 

2. 커밋을 컴포넌트 단위로 더 쪼개면 좋을 것 같습니다

who0803 commented:
> 스토리북 파일까지 꼼꼼하게 만들어주셨네요 👍 👍
> 
> 질문 몇 개 추가합니다!
> 
> 1. Auth폴더 내부의 Input, Textarea, ErrorMessage 컴포넌트
> 
> * Auth관련한 곳에서만 사용할 컴포넌트인가요? 구현 내용을 보면 공통 컴포넌트에 가까워보여서 질문드립니다
> * getFieldProps 반환값이 아닌 함수 자체를 props로 받아오는 이유가 무엇인가요?
> 
> 2. 커밋을 컴포넌트 단위로 더 쪼개면 좋을 것 같습니다

여기에 대한 답으로써

1. 일단은 common 폴더의 컴포넌트와 이름이 겹치고, common에 있는 컴포넌트를 바로 대체하기에는 다른 곳에서 쓰고있는 곳이 많아서 임시로 Auth 폴더에 위치시켰습니다. 제가 pr 날릴 때 미리 설명한다는 게 까먹었네요
2. getFieldProps 반환값이 4개나 되기 때문에 Input 컴포넌트에 공통적으로 꼭 들어가야하는 값들을 묶어서 내보내기 위해서 했습니다.
3. 죄송합니다. pr단위를 어떻게 나눌까 생각하다가 제가 생각한 pr 단위 그대로 커밋해버렸습니다... 반성합니다...

who0803 committed:
fix: Fix optional chaining logic in className

옵셔널 체이닝이 중첩되어 가독성이 떨어지는 부분 수정

who0803 committed:
feat: Fix children prop type in Input component

인풋 컴포넌트의 children prop의 타입 수정

who0803 committed:
feat: Create Message component for atomic component

에러메세지를 범용적으로 사용하기 위해서 아토믹한 메세지 컴포넌트 구현

who0803 committed:
fix: remove serverError logic in ErrorMessage component

기존 ErrorMessage 컴포넌트에 있던 serverError 로직을 삭제하여
ErrorMessage 컴포넌트는 validation을 위한 용도로만 사용되도록 수정
serverError 로직은 아토믹한 Message 컴포넌트를 만들어서 해결

해당 컴포넌트 수정에 따른 스토리북, 스타일 수정

who0803 committed:
fix: Fix storybook of Form component

ErrorMessage 컴포넌트 수정과 Message 컴포넌트 생성에 따른 스토리북 수정

CHAERYEON-LEE reviewed (commented):
넘모 좋습니다

who0803 committed:
fix: remove lg2 prop in Input component

변경 전
- 기존에는 lg2 prop을 통해서 label의 sr-only 적용 유무와 border-radius
  스타일을 적용했다.

변경 후
- lg2 prop을 제거하고 border prop을 만들어 Input 컴포넌트의
  border-radius를 적용
- placeholder prop의 유무에 따라서 label의 sr-only 적용 결정

who0803 committed:
fix: Change name of ErrorMessage Component to ValidationErrorMessage

- ErrorMessage 컴포넌트의 이름을 ValidationErrorMessage 컴포넌트로 수정
- Input의 스토리북에서 ErrorMessage를 import해오던 것을 바뀐이름으로
  수정

who0803 committed:
fix: Change name of Message component to ErrorMessage

- Message 컴포넌트의 이름을 ErrorMessage로 변경
- 이에 따라 import 수정

who0803 committed:
fix: Fix storybook of Form and Textarea component

- 컴포넌트 이름 변경 및 prop 수정이로 인한 Form, Textarea 컴포넌트의 스토리북 수정
